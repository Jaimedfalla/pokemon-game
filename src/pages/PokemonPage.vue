<template>
  <h1 v-if="!pokemon">Loading, please wait...</h1>
  <div v-else>
    <h1>Which pokemon is it?</h1>
    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"
    ></PokemonPicture>
    <PokemonOptions
      :optionsPokemon="options"
      @selection="selectPokemon"
    ></PokemonOptions>
    <template v-if="showAnswer">
      <h2 class="fade-in" :class="[isCorrect ? 'correct' : 'error']">
        {{ message }}
      </h2>
      <button @click="restartGame">Restart Game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/PokemonOptions";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      options: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
      isCorrect: false,
    };
  },
  methods: {
    async mixPokemonsArray() {
      this.options = await getPokemonOptions();
      const id = Math.floor(Math.random() * 4);
      this.pokemon = this.options[id];
    },
    selectPokemon(id) {
      this.showPokemon = true;
      this.showAnswer = true;
      this.isCorrect = this.pokemon.id === id;
      this.message = `${
        this.isCorrect ? "Correct!!! :D:D:D" : "Incorrect"
      }, it is ${this.pokemon.name}`;
    },
    async restartGame() {
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      await this.mixPokemonsArray();
    },
  },
  mounted() {
    this.mixPokemonsArray();
  },
};
</script>

<style scoped>
.error {
  color: #d43737;
}

.correct {
  color: #86d437;
}
</style>
