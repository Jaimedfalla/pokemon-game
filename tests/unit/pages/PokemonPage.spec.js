import { shallowMount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage.vue";
import { pokemons } from "../mocks/pokemons.mock";

describe("Pokemon page component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          options: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          isCorrect: false,
        };
      },
    });
  });

  test("Shoul call mixPokemonArray method when mounting", () => {
    const mixPokemon = jest.spyOn(PokemonPage.methods, "mixPokemonsArray");
    shallowMount(PokemonPage);
    expect(mixPokemon).toHaveBeenCalled();
  });

  test("Should match with snapshot when pokemons have loaded", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Should show Pokemon Picture and Pokemon Options components", () => {
    const picture = wrapper.find("pokemon-picture-stub");
    const options = wrapper.find("pokemon-options-stub");

    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();
    expect(picture.attributes("pokemonid")).toBe(pokemons[0].id.toString());
    console.log(options.attributes("optionspokemon"));
    expect(options.attributes("optionspokemon")).toBeTruthy();
  });

  test("Should call check answer method", async () => {
    await wrapper.vm.selectPokemon(pokemons[0].id);
    const h2 = wrapper.find("h2");
    expect(h2.exists()).toBeTruthy();
    expect(h2.text()).toContain("Correct");
    expect(wrapper.vm.isCorrect).toBe(true);
  });
});
