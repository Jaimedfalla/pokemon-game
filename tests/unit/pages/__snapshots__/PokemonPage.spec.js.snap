// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon page component Should match with snapshot when pokemons have loaded 1`] = `
<div>
  <h1>Which pokemon is it?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub optionspokemon="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
